### PR TITLE
Support rollback nodes in scenario service

### PR DESCRIPTION
**Pull request by cocreature**
This focuses on the semantics rather than the display in Daml Studio
which needs more work (and seems not all that important at this
stage).

This already uncovered a bug which also applies outside of scenarios:

The consumedBy field was not affected by rollbacks which breaks the
mustBeActive check in partial transactions. This PR fixes this by
caching on try and restoring on rollback.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/NodeInfo.scala [hunk 1]
```diff
@@ -89,11 +89,11 @@ object NodeInfo {
 
     final def requiredAuthorizers: Set[Party] =
       // TODO https://github.com/digital-asset/daml/issues/8020
-      sys.error("rollback nodes are not supported")
+      Set.empty
 
     final def informeesOfNode: Set[Party] =
       // TODO https://github.com/digital-asset/daml/issues/8020
-      sys.error("rollback nodes are not supported")
+      Set.empty
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/NodeInfo.scala:96]: This is obviously broken but it is good enough to fix the scenario service (who doesn’t use those methods). I would propose to merge this and then fix it in a separate PR.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala [hunk 2]
```diff
@@ -427,9 +427,11 @@ object ScenarioLedger {
...
 
                   node match {
-                    case _: NodeRollback[_] =>
-                      // TODO https://github.com/digital-asset/daml/issues/8020
-                      sys.error("rollback nodes are not supported")
+                    case rollback: NodeRollback[_] =>
+                      processNodes(
+                        Right(newCache),
+                        (Some(nodeId) -> rollback.children.toList) :: idsToProcess,
+                      )
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala:434]: This is completely broken, will spend some more time thinking about it.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/ledger/BlindingTransaction.scala [hunk 3]
```diff
@@ -58,48 +58,53 @@ object BlindingTransaction {
         parentExerciseWitnesses: Set[Party],
         nodeId: NodeId,
     ): BlindState = {
-      val action = tx.nodes.get(nodeId) match {
+      tx.nodes.get(nodeId) match {
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/ledger/BlindingTransaction.scala:61]: Sorry I shouldn’t be sneaking this in here. Will try to spin it out into a separate PR with tests.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 4]
```diff
@@ -549,6 +555,8 @@ private[lf] case class PartialTransaction(
         copy(
           context = info.parent.addRollbackChild(info.nodeId, context.nextActionChildIdx),
           nodes = nodes.updated(info.nodeId, rollbackNode),
+          // Undo any consuming choices in the rollback.
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:558]: Same here, will also try to get this into a separate PR with a standalone test.

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:558]: https://github.com/digital-asset/daml/pull/9430

**Review by sofiafaro-da — COMMENTED** [review]
Tests look good to me. I can't comment on the scala side.

File: compiler/damlc/tests/daml-test-files/ExceptionSemantics.daml [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+divulgence = scenario do
+  p1 <- getParty "p1"
+  p2 <- getParty "p2"
+  cid <- submit p1 $ create (K p1 0)
+  divulger <- submit p2 $ create (Fetcher p2 p1)
+  fetcher <- submit p1 $ create (Fetcher p1 p2)
+  submitMustFail p2 $ exercise fetcher (Fetch cid)
+  submit p1 $ exercise divulger (RollbackFetch cid)
+  submit p2 $ exercise fetcher (Fetch cid)
```
Comment (sofiafaro-da) [inline compiler/damlc/tests/daml-test-files/ExceptionSemantics.daml:133]: Interesting case!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/ledger/BlindingTransaction.scala (hunk 3 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/ledger/BlindingTransaction.scala:61]: https://github.com/digital-asset/daml/pull/9431

**Comment by cocreature** [comment]
Dependent PRs are all in so should be ready now.

**Review by nickchapman-da — APPROVED** [review]
LGTM
The additions to the `unwind()` code in Speedy.scala look fine.
Nice catch!
